### PR TITLE
ramips: mt7621: add migration script for WLAN PCI paths

### LINKED ISSUE #9374: mt7621: R6220 2.4GHz radio0 issues
Hello, i have issues with 2.4ghz wireless not working on openwrt 21.2.02 on my Netgear R6220, i have also tested snapshot build, with same result.
Im pretty sure that issue is caused by bad eraseblocks, becouse every (x) reboots radio0 would be up and working util next "power outage"

```
root@OpenWrt:~# dmesg | grep eraseblock
[    1.922023] Bad eraseblock 847 at 0x0000069e0000
```
image

I have read on forum that kernel 4.x had applied patch and 5.x wont.

I tried to install openwrt 19.07.9 whih i guess fixed my problem (gonna test it longer) But thats not a good solution to me to stick with old firmware. 

Also, i have been trying to move partition "factory" to other offsets using imagebuilder, but seems i dont know how, or it need be rebuilded from sourcecode, becouse nothing in `/prod/mtd` is changing when builded new image.

Ps. Sorry for my bad english (not my main lang)

### PR DESCRIPTION
**Pull request by Djfe**
PCI paths of the WLAN devices have changed between kernel 5.10 and 5.15; migrate config so existing wifi-iface definitions don't break.

This is implemented as a hotplug handler rather than a uci-defaults script as the migration script must run before the 10-wifi-detect hotplug handler.

based on b452af23a8602ebf7bfb0eb084383ecd595face5

migration was forgotten when device trees were adjusted in
688697889c9ceb9d2d8d413d1abcc5fa7edb402b
c77913be5be5f11a0f20dba1b45b96470099e8a8

fixes #9374

affected devices:
Netgear R6220
Netgear WAC104
Netgear WNDR3700 v5
Zbtlink ZBT-WE1326
Wiflyer WF3526-P
Arcadyan WE420223-99
Beeline Smartbox Flash (Arcadyan WG443223)
MTS WG430223 (Arcadyan WG430223)

___________________

Users of the affected devices noticed in https://github.com/openwrt/openwrt/issues/9374#issuecomment-1457297507 that we need to migrate the wifi paths still. @DragonBluep 

This patch is solely based on https://github.com/openwrt/openwrt/commit/b452af23a8602ebf7bfb0eb084383ecd595face5
I checked the code. If I understand it correctly it should be run once only, even on the Zbtlink/Wiflyer device, as it checks first, whether the migrated new "to" path already exists :)
So even though one path keeps on existing it will migrate the wifi-device successfully to the new paths.

If you want me to add a code comment to the top of the file, then let me know :)
~~This code wasn't tested, yet.~~
I don't own any of the affected hardware.
The PCI address of the zbtlink wifi-devices is partially a guess and based on the Netgear addresses.

EDIT: The code was tested successfully by @mbaumga on one of the affected Netgear devices. 
The first force push added `Tested-by` to the commit message
Now I'm only waiting for a successful test of the affected Zbt device :)

The second force push added Arcadyan devices upon request of @csharper2005 
The third force push addressed the changes @csharper2005 requested and adjusted the pci paths of the Zbt-based devices slightly.
The fourth force push added Tested-by @csharper2005

### CODE REVIEW

**Comment by csharper2005** [comment]
@Djfe, nice work. Some Arcadyan devices are also need pci paths migration. 

And I have a question. Do we need a similar script if we make a downgrade? For example, from master to 22.03.x.

**Comment by Djfe** [comment]
> @Djfe, nice work. Some Arcadyan devices are also need pci paths migration.

Lantiq target? any hints which ones would be helpful
> And I have a question. Do we need a similar script if we make a downgrade? For example, from master to 22.03.x.

One of the other scripts for wifi migration I stumbled upon does this ([See](https://github.com/openwrt/openwrt/blob/ec6293febc244d187e71a6e54f44920be679cde4/target/linux/mvebu/cortexa9/base-files/etc/hotplug.d/ieee80211/00-wifi-config-migrate)), but:
That only works if the patch gets backported and if another release of OpenWRT 22 is planned. (downgrade to 22.03.03 will never work regardless what I do here)

Since the Zbt device requires special handling (one path is kept but moved between radio0 and radio1), I like how neoraider implemented this script and would like to keep it failsafe.
If you want such a script for Release 22, then I'd open a new PR against branch 22 that reverses the change.

**Comment by Djfe** [comment]
ah, I see #12091

Usually I wouldn't bother since it was just added in Snapshot, but adding this to this patch is no effort either, so I'll do it :)

The thing is: this can't fix your wireless config, if you already upgraded to a snapshot on Kernel 5.15. This needs to be run before the new slot is detected as a new (third) radio by a hotplug.d script.
If you already have 3 radios, then you are only left with resetting your wireless config.

The output of cat `/etc/config/wireless` would be helpful so I can be sure, the pcie path is correct.

**Comment by Djfe** [comment]
Ok, it's three Arcadyan devices, I think I found all of them :)
by now I'm fairly certain that the last part of the pci address `0000:02:00.0` might be different for Zbt and for the Arcadyan devices and I need people who own these devices to confirm, what these paths actually are.

**Comment by csharper2005** [comment]
@Djfe this is the block for ramips Arcadyan. They have DBDC wireless, so 2 lines here.
```
arcadyan,we420223-99|\
beeline,smartbox-flash|\
mts,wg430223)
	migrate_radio '1e140000.pcie/pci0000:00/0000:00:00.0/0000:01:00.0' '1e140000.pcie/pci0000:00/0000:00:01.0/0000:01:00.0'
	migrate_radio '1e140000.pcie/pci0000:00/0000:00:00.0/0000:01:00.0+1' '1e140000.pcie/pci0000:00/0000:00:01.0/0000:01:00.0+1'
	;;
```
[mts,wg430223]
Tested-by: Mikhail Zhilkin <csharper2005@gmail.com>

File: target/linux/ramips/mt7621/base-files/etc/hotplug.d/ieee80211/05-wifi-migrate [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+netgear,r6220|\
+netgear,wac104|\
+netgear,wndr3700-v5)
+	migrate_radio '1e140000.pcie/pci0000:00/0000:00:01.0/0000:02:00.0' '1e140000.pcie/pci0000:00/0000:00:02.0/0000:02:00.0'
+	;;
+zbtlink,zbt-we1326)
+	migrate_radio '1e140000.pcie/pci0000:00/0000:00:01.0/0000:02:00.0' '1e140000.pcie/pci0000:00/0000:00:02.0/0000:02:00.0'
+	migrate_radio '1e140000.pcie/pci0000:00/0000:00:00.0/0000:02:00.0' '1e140000.pcie/pci0000:00/0000:00:01.0/0000:02:00.0'
+	;;
+arcadyan,we420223-99|\
```
Comment (csharper2005) [inline target/linux/ramips/mt7621/base-files/etc/hotplug.d/ieee80211/05-wifi-migrate:64]: I would suggest to move this block at the top (according to the alphabetical order).

File: target/linux/ramips/mt7621/base-files/etc/hotplug.d/ieee80211/05-wifi-migrate [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+beeline,smartbox-flash|\
+mts,wg430223)
+	migrate_radio '1e140000.pcie/pci0000:00/0000:00:00.0/0000:02:00.0' '1e140000.pcie/pci0000:00/0000:00:01.0/0000:02:00.0'
+	;;
+
+esac
+
+$WIRELESS_CHANGED && uci commit wireless
+
+exit 0
```
Comment (csharper2005) [inline target/linux/ramips/mt7621/base-files/etc/hotplug.d/ieee80211/05-wifi-migrate:74]: Could you add a newline, please?

**Comment by Djfe** [comment]
> @Djfe this is the block for ramips Arcadyan. They have DBDC wireless, so 2 lines here.

Are you certain, that we need to migrate two paths?
I thought only the 5GHz wifi path changed with the kernel upgrade  (only one)

**Comment by Djfe** [comment]
ah nvm, as you said DBDC, so one pcie slot in use only. next force push incoming addressing your changes

**Comment by csharper2005** [comment]
> Are you certain, that we need to migrate two paths?

Yes, I'm sure.
```
root@wg430223:~# cat /etc/config/wireless | grep path
        option path '1e140000.pcie/pci0000:00/0000:00:01.0/0000:01:00.0'
        option path '1e140000.pcie/pci0000:00/0000:00:01.0/0000:01:00.0+1'
```

**Comment by csharper2005** [comment]
> next force push incoming addressing the changes you requested

Thanks a lot!

**Comment by Djfe** [comment]
Thanks for testing this on your hardware :)
This remains a draft until, we have confirmation that the paths are correct for the Zbt devices.

**Comment by Djfe** [comment]
@dfateyev do you have time to test this on your Wiflyer WF3526-P maybe?
What's the output of `cat /etc/config/wireless` on Release and Master with this patch applied? (keep config when upgrading from release to master)

**Comment by dfateyev** [comment]
> @dfateyev do you have time to test this on your Wiflyer WF3526-P maybe?

I can't upgrade this device from 19.07 to newer version now: it's in use, and has [stability issues](https://github.com/openwrt/mt76/issues/719) with MT7603E on newer versions. If it can help, there is output from the 19.07.10, installed from scratch:

```
# cat /etc/config/wireless | grep path
	option path 'pci0000:00/0000:00:01.0/0000:02:00.0'      <<<  2.4GHz
	option path 'pci0000:00/0000:00:00.0/0000:01:00.0'      <<<  5Ghz

```

**Comment by Djfe** [comment]
understandable.
That actually helps a lot and confirmed my suspictions. The changes I made to the Zbt pci paths in force-push seem to be accurate.

I feel like this is ready to be reviewed now.
The reviewers may decide whether it's also ready to be merged.

**Pull request merged**